planner/funcdep/fast_int_set_test.go

```diff
@@ -0,0 +1,515 @@
+		if _, ok := is[i]; ok {
+			delete(is, i)
+		}
+	}
+}
+
+func (is IntSet) Difference2(target1, target2 IntSet) {
+	for i := range target1 {
+		if _, ok := target2[i]; ok {
+			delete(is, i)
+		} else {
+			is[i] = struct{}{}
+		}
+	}
+}
+
+// Union is used to union the IntSet itself with others
+func (is IntSet) Union(target IntSet) {
+	// deduplicate
+	for i := range target {
+		if _, ok := is[i]; ok {
+			continue
+		}
+		is[i] = struct{}{}
+	}
+}
+
+// Equals is used to judge whether two IntSet are semantically equal.
+func (is IntSet) Equals(target IntSet) bool {
+	if len(is) != len(target) {
+		return false
+	}
+	for i := range target {
+		if _, ok := is[i]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func (is *IntSet) CopyFrom(target IntSet) {
+	*is = NewIntSetWithCap(len(target))
+	for k, v := range target {
+		(*is)[k] = v
+	}
+}
+
+func (is IntSet) SortedArray() []int {
+	arr := make([]int, 0, len(is))
+	for k := range is {
+		arr = append(arr, k)
+	}
+	sort.Slice(arr, func(i, j int) bool { return arr[i] < arr[j] })
+	return arr
+}
+
+func (is IntSet) Insert(k int) {
+	is[k] = struct{}{}
+}
+
+func NewIntSet() IntSet {
+	return make(map[int]struct{})
+}
+
+func NewIntSetWithCap(cap int) IntSet {
+	return make(map[int]struct{}, cap)
+}
+
+func TestFastIntSetBasic(t *testing.T) {
+	ass := assert.New(t)
+
+	// Test Insert, Remove, Len, Has.
+	fis := FastIntSet{}
+	fis.Insert(1)
+	fis.Insert(2)
+	fis.Insert(3)
+	ass.Equal(fis.Len(), 3)
+	ass.True(fis.Has(1))
+	ass.True(fis.Has(2))
+	ass.True(fis.Has(3))
+	fis.Remove(2)
+	ass.Equal(fis.Len(), 2)
+	ass.True(fis.Has(1))
+	ass.True(fis.Has(3))
+	fis.Remove(3)
+	ass.Equal(fis.Len(), 1)
+	ass.True(fis.Has(1))
+	fis.Remove(1)
+	ass.Equal(fis.Len(), 0)
+
+	// Test Next (only seek non-neg)
+	fis.Insert(6)
+	fis.Insert(3)
+	fis.Insert(0)
+	fis.Insert(-1)
+	fis.Insert(77)
+	n, ok := fis.Next(intsets.MinInt)
+	ass.True(ok)
+	ass.Equal(n, 0)
+	n, ok = fis.Next(n + 1)
+	ass.True(ok)
+	ass.Equal(n, 3)
+	n, ok = fis.Next(n + 1)
+	ass.True(ok)
+	ass.Equal(n, 6)
+	n, ok = fis.Next(n + 1)
+	ass.True(ok)
+	ass.Equal(n, 77)
+	n, ok = fis.Next(n + 1)
+	ass.False(ok)
+
+	// Test Clear and IsEmpty.
+	fis.Clear()
+	ass.Equal(fis.Len(), 0)
+	ass.True(fis.IsEmpty())
+
+	// Test ForEach (seek all) and SortedArray.
+	fis.Insert(1)
+	fis.Insert(-1)
+	fis.Insert(77)
+	var res []int
+	fis.ForEach(func(i int) {
+		res = append(res, i)
+	})
+	res1 := fis.SortedArray()
+	ass.Equal(len(res), 3)
+	ass.Equal(len(res1), 3)
+	ass.Equal(res, res1)
+
+	// Test Copy,  CopyFrom and Equal
+	cp := fis.Copy()
+	ass.Equal(fis.Len(), cp.Len())
+	ass.Equal(fis.SortedArray(), cp.SortedArray())
+	ass.True(fis.Equals(cp))
+
+	cpf := FastIntSet{}
+	intervene := 100
+	cpf.Insert(intervene)
+	cpf.CopyFrom(fis)
+	ass.Equal(cpf.Len(), cp.Len())
+	ass.Equal(cpf.SortedArray(), cp.SortedArray())
+	ass.True(cpf.Equals(cp))
+}
+
+// A Mutex is a mutual exclusion lock.
+type Mutex struct {
+	sync.Mutex
+}
+
+var mtx Mutex
```

Comment:
Our tests are running without parallel mode. they are isolated. so this Mutex is useless.

Reply:
done